### PR TITLE
Serialize RSA public keys

### PR DESCRIPTION
**Pull request by reaperhulk**
Outstanding work:
- [x] Discuss `PublicFormat` naming.
- [x] Implement "bare" public key support (refs #1705)
- [x] rebase after #1711 lands

### CODE REVIEW

**Comment by alex** [comment]
You can rebase all teh stuff out of this now.

File: docs/hazmat/primitives/asymmetric/serialization.rst [hunk 1]
```diff
@@ -309,6 +309,25 @@ Serialization Formats
...
 
+.. class:: PublicFormat
+
+    .. versionadded:: 0.8
+
+    An enumeration for public key formats. Used with the ``public_bytes``
+    method available on
+    :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKeyWithSerialization`.
+
+    .. attribute:: SubjectPublicKeyInfo
```
Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: Is this the best name for this?

File: docs/hazmat/primitives/asymmetric/serialization.rst [hunk 2]
```diff
@@ -309,6 +309,25 @@ Serialization Formats
...
+
+        This is the typical public key format. It consists of an algorithm
+        identifier and the public key as a bit string. Choose this unless
+        you have specific needs.
+
+    .. attribute:: PKCS1
+
+        Just the public key elements (without the algorithm identifier). This
+        format is occasionally used by various systems.
```
Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: This is also only applicable for RSA keys?

**Comment by alex** [comment]
Besides those comments this looks fan-dab-tastic

File: docs/hazmat/primitives/asymmetric/serialization.rst (hunk 2 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: Pretty much. Probably worth documenting that explicitly for now (DSA does exist but OpenSSL doesn't expose it as a macro so by default you can't load it unless you use some undocumented functions).

Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: Let's add that to the docs then.

On Sun, Mar 8, 2015 at 3:08 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/serialization.rst
> <https://github.com/pyca/cryptography/pull/1706#discussion_r26007880>:
>
> > +    .. versionadded:: 0.8
> > +
> > +    An enumeration for public key formats. Used with the ``public_bytes``
> > +    method available on
> > +    :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKeyWithSerialization`.
> > +
> > +    .. attribute:: SubjectPublicKeyInfo
> > +
> > +        This is the typical public key format. It consists of an algorithm
> > +        identifier and the public key as a bit string. Choose this unless
> > +        you have specific needs.
> > +
> > +    .. attribute:: PKCS1
> > +
> > +        Just the public key elements (without the algorithm identifier). This
> > +        format is occasionally used by various systems.
>
> Pretty much. Probably worth documenting that explicitly for now (DSA does
> exist but OpenSSL doesn't expose it as a macro so by default you can't load
> it unless you use some undocumented functions).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1706/files#r26007880>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: docs/hazmat/primitives/asymmetric/serialization.rst (hunk 1 above)
Comment (reaperhulk) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: It's the name used by PyCrypto and [Crypto++](http://www.cryptopp.com/wiki/Keys_and_Formats#Public_Key_Format). Go calls it a "DER-encoded PKIX structure" but if you look in the source code it says "pkixPublicKey reflects a PKIX public key structure. See SubjectPublicKeyInfo in RFC 3280."

Comment (alex) [inline docs/hazmat/primitives/asymmetric/serialization.rst]: Ok. Good enough for me.

On Sun, Mar 8, 2015 at 3:12 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/serialization.rst
> <https://github.com/pyca/cryptography/pull/1706#discussion_r26007907>:
>
> > @@ -309,6 +309,25 @@ Serialization Formats
> >          encryption. Choose this unless you have explicit legacy compatibility
> >          requirements.
> >
> > +.. class:: PublicFormat
> > +
> > +    .. versionadded:: 0.8
> > +
> > +    An enumeration for public key formats. Used with the ``public_bytes``
> > +    method available on
> > +    :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPublicKeyWithSerialization`.
> > +
> > +    .. attribute:: SubjectPublicKeyInfo
>
> It's the name used by PyCrypto and Crypto++
> <http://www.cryptopp.com/wiki/Keys_and_Formats#Public_Key_Format>. Go
> calls it a "DER-encoded PKIX structure" but if you look in the source code
> it says "pkixPublicKey reflects a PKIX public key structure. See
> SubjectPublicKeyInfo in RFC 3280."
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1706/files#r26007907>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Pull request merged**